mflike/foreground.py

```diff
@@ -482,36 +640,271 @@ def _bandpass_construction(self, _initialize=False, **params):
+
+                        if "tt" in self.requested_cls or "te" in self.requested_cls:
+                            trans_normT = trapezoid(
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT, nub, axis=0
+                            )
+                            ratioT = (
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blT / trans_normT
+                            )
+                            self.bandint_freqs_T.append([nub, ratioT])
+
+                        if "te" in self.requested_cls or "ee" in self.requested_cls:
+                            trans_normP = trapezoid(
+                                bp[..., np.newaxis] *  _cmb2bb(nub)[..., np.newaxis] * blP, nub, axis=0
+                            )
+                            ratioP = (
+                                bp[..., np.newaxis] * _cmb2bb(nub)[..., np.newaxis] * blP / trans_normP
+                            )
+                            self.bandint_freqs_P.append([nub, ratioP])
 
         # fgspectra can't mix monofrequency with [nu, bp]. If one channel is mono-frequency then we
         # assume all of them and pass to fgspectra an array (not list!!) of frequencies
         if data_are_monofreq:
-            self.bandint_freqs = np.asarray(self.bandint_freqs)
+            self.bandint_freqs_T = np.asarray(self.bandint_freqs_T)
+            self.bandint_freqs_P = np.asarray(self.bandint_freqs_P)
             if self._initialized:
                 self.log.info("bandpass is delta function, no band integration performed")
         self._initialized = True
 
+    ###########################################################################
+    ## This part deals with beam functions, i.e. reading beam from file or
+    ## computing it as a Gaussian beam. We also have a function to compute
+    ## the correction expected for a Gaussian beam in case of bandpass shift
+    ## that should be applied to any beam profile
+    ###########################################################################
+    def _read_yaml_file(self, file_path):
+        
+        from cobaya.yaml import yaml_load_file
+        
+        filename = "%s.yaml" % file_path
+        file = yaml_load_file(file_name = filename)
+
+        return file
+
+    def _init_beam_from_file(self):
+        """
+        Reads the beam profile from an external file or the sacc file.
+        It has to be a dictionary ``{"{exp}_s0": {"nu": nu, "beams": array(freqs, ells+2)},
+        "{exp}_s2": {"nu": nu, "beams": array(freqs, ells+2)},...}``
+        including temperature and polarization beams.
+        """
+
+        if not self.beam_file:
+            # option to read beam from sacc
+            if not bool(self.beams):
+                raise ValueError("Beams not stored in sacc files!")
+        else:
+            self.beams = self._read_yaml_file(self.beam_file)
```

Comment:
So `self.beam_file` also does not have file extension ?

Reply:
Now I have fixed it, the code assumes the extension is included